File: auth/src/models/module.rs
use chrono::{serde::ts_seconds, DateTime, Utc};
use serde::Serialize;

use crate::types;

#[derive(Clone, Serialize, Debug, sqlx::FromRow)]
pub struct ModuleModel {
    pub id: i32,
    pub serial_number: String,
    pub private_key: String,

    #[serde(with = "ts_seconds")]
    pub created_at: DateTime<Utc>,
}

impl ModuleModel {
    pub async fn get_by_serial_number(
        context: types::AppContext,
        serial_number: String,
    ) -> Result<ModuleModel, sqlx::Error> {
        let mut conn = context.database_pool.acquire().await?;

        sqlx::query_as!(
            ModuleModel,
            "SELECT *
            FROM modules
            WHERE serial_number = $1;",
            serial_number
        )
        .fetch_one(&mut conn)
        .await
    }

    pub async fn create(
        context: types::AppContext,
        serial_number: String,
        private_key_hash: String,
    ) -> Result<(), sqlx::Error> {
        let mut conn = context.database_pool.acquire().await?;

        let result = sqlx::query!(
            "INSERT INTO modules
            (serial_number, private_key)
            VALUES ($1, $2);",
            serial_number,
            private_key_hash
        )
        .execute(&mut conn)
        .await?;

        Ok(())
    }
}

File: go-auth/repositories/module.go
package repositories

import (
	"time"

	"github.com/jmoiron/sqlx"
)

type PostgresAuthModule struct {
	Id           int       `db:"id"`
	SerialNumber string    `db:"serial_number"`
	PrivateKey   string    `db:"private_key"`
	CreatedAt    time.Time `db:"created_at"`
	UpdatedAt    time.Time `db:"updated_at"`
}

type AuthModuleRepository struct {
	db *sqlx.DB
}

func NewModuleRepository(db *sqlx.DB) *AuthModuleRepository {
	return &AuthModuleRepository{db}
}

func (pmr *AuthModuleRepository) Create(serial_number string, private_key string) error {
	create_module_query := `
		INSERT INTO modules (serial_number, private_key)
		VALUES ($1, $2);
	`

	_, err := pmr.db.Exec(create_module_query, serial_number, private_key)
	if err != nil {
		return err
	}

	return nil
}

func (pmr *AuthModuleRepository) GetBySerialNumber(serial_number string) (*PostgresAuthModule, error) {
	get_module_query := `
 		SELECT *
		FROM modules
		WHERE serial_number = $1;
	`

	row := pmr.db.QueryRowx(get_module_query, serial_number)

	var pm PostgresAuthModule
	err := row.StructScan(&pm)
	if err != nil {
		return nil, err
	}

	return &pm, nil
}
